store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/retriever/HBaseRetriever.java

```diff
@@ -0,0 +1,231 @@
+import uk.gov.gchq.gaffer.store.StoreException;
+import uk.gov.gchq.gaffer.user.User;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class HBaseRetriever<OP extends Output<CloseableIterable<? extends Element>> & GraphFilters & Options> implements CloseableIterable<Element> {
+    private final ElementSerialisation serialisation;
+    private final RowRangeFactory rowRangeFactory;
+    private final ElementValidator validator;
+    private final Iterable<? extends ElementId> ids;
+    private final HBaseStore store;
+    private final Authorizations authorisations;
+    private final OP operation;
+    private final byte[] extraProcessors;
+
+    private CloseableIterator<Element> iterator;
+
+    public HBaseRetriever(final HBaseStore store,
+                          final OP operation,
+                          final User user,
+                          final Iterable<? extends ElementId> ids,
+                          final Class<?>... extraProcessors) throws StoreException {
+        this.serialisation = new ElementSerialisation(store.getSchema());
+        this.rowRangeFactory = new RowRangeFactory(serialisation);
+        this.validator = new ElementValidator(operation.getView());
+        this.store = store;
+        this.operation = operation;
+        this.ids = ids;
+        if (null != user && null != user.getDataAuths()) {
+            this.authorisations = new Authorizations(
+                    user.getDataAuths().toArray(new String[user.getDataAuths().size()]));
+        } else {
+            this.authorisations = new Authorizations();
+        }
+
+        if (null != extraProcessors && extraProcessors.length > 0) {
+            this.extraProcessors = StringUtil.toCsv(extraProcessors);
+        } else {
+            this.extraProcessors = null;
+        }
+    }
+
+    @Override
+    public void close() {
+        if (iterator != null) {
+            iterator.close();
+            iterator = null;
+        }
+    }
+
+    @Override
+    public CloseableIterator<Element> iterator() {
+        // By design, only 1 iterator can be open at a time
+        close();
+
+        final ResultScanner scanner = getScanner();
+        if (null != scanner) {
+            iterator = new HBaseRetrieverIterable(scanner).iterator();
+        } else {
+            iterator = new WrappedCloseableIterator<>(Collections.emptyIterator());
+        }
+
+        return iterator;
+    }
+
+    private ResultScanner getScanner() {
+        Table table = null;
+        try {
+            final Scan scan = new Scan();
+
+            if (null != ids) {
+                final List<MultiRowRangeFilter.RowRange> rowRanges = new ArrayList<>();
+                for (final ElementId id : ids) {
+                    rowRanges.addAll(rowRangeFactory.getRowRange(id, operation));
+                }
+
```

Comment:
In the AccumuloStore this is done in batches. There's nothing to say that the iterable of ids fits into memory. The API is designed to allow querying for arbitrarily large iterables of ids, returning arbitrarily large numbers of results.

Happy for this to stay as it is for now, and for this improvement to be suggested in an issue.